src/averageAges.js

```diff
@@ -55,10 +78,35 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  if (onlyWithSon !== undefined) {
+    const mothers = people.filter(person =>
+      people.some(p => person.mother === p.name) && person.sex === 'm'
+    );
+    const mothersAndChildDiff = mothers.map(child => {
+      const mother = people.find(person => person.name === child.mother);
+
+      if (mother) {
+        return child.born - mother.born;
+      }
+    });
+    const sumAge = mothersAndChildDiff.reduce((acc, current) => acc + current);
+
+    return sumAge / mothersAndChildDiff.length;
+  } else {
+    const mothers = people.filter(person =>
+      people.some(p => person.mother === p.name)
+    );
+    const mothersAndChildDiff = mothers.map(child => {
+      const mother = people.find(person => person.name === child.mother);
+
+      if (mother) {
+        return child.born - mother.born;
+      }
+    });
+    const sumAge = mothersAndChildDiff.reduce((acc, current) => acc + current);
+
+    return sumAge / mothersAndChildDiff.length;
+  }
```

Comment:
There is a lot of duplication of the same code here, try to simplify it and not explicitly check for undefined